### ISSUE TITLE
[Doc]: Revisit interface terminology: Axes interface / pyplot interface

### ISSUE THREAD

**Issue by timhoffm** [matplotlib/matplotlib#26388]
### Documentation Link

https://matplotlib.org/devdocs/api/index.html

### Problem

We currently use the terms the *explicit API* and *implicit API*. Even though they were the result of lengthy discussions, I still have a hard time using them. I bet that most of our users do not know these terms.  And also when given these terms, I suppose most would not know what they mean. In my experience, I always have to follow up on these terms with an explanatory block.

I think the problem is that they are very technical and not something a user can relate to intuitively: Explicit/implicit is the mode of operation; i.e. how the code they write reference the Figure and Axes. This is not a primary concern of users.

#### Use *interface* instead of *API*.

API is too much on the technical side (which objects/functions/methods are there and what are their signatures). To me *interface* has more emphasis on the conceptual aspect (how to I work with this). Also, *interface* can have a

#### Distinguishing characteristics

Use the the things the user interacts with as the distinguishing characteristic:

- <s>Variant 1) *conceptual distinction*: **the object interface / the function interface**</s>
- Variant 2) *practical distinction*: **the Axes interface / the pyplot interface**

I propose to go with variant 2). While object/function is more concrete than explicit/implicit, it's still somewhat generic (what objects/functions?). Also, the OOP aspect is dilluted because you usually do not instantiate objects yourself and the interactions is mostly limited to Figure and Axes objects.

In variant 2) one could have gone with "the Figure-and-Axes interface", but that seems too verbose, and given that operations on the Figure are much less frequent compared to operations on the Axes, I think we can afford that simplification. The nice thing about this description is that it directly relates to the prefixes that are used for plotting: Axes -> `ax`; pyplot --> `plt`.

```
# The Axes interface
fig, ax = plt.subplots()
ax.plot(x, y)
ax.set_xlabel(...)
ax.set_ylabel(...)
ax.set_title(...)

# The pyplot interface
plt.plot(x, y)
plt.xlabel(...)
plt.ylabel(...)
plt.title(...)
```

**Comment by jklymak** [comment]
Agreed - implicit and explicit is hard to keep track of.  OTOH, it could still be mentioned.

**Comment by jklymak** [comment]
You can rephrase it if you like, but it is in contrast to "third-party" interfaces in the next section.

**Comment by tacaswell** [comment]
A disadvantage of "axes interface" vs "pyplot interface" is that it does not actually convey any information about what they are or why you would want to pick one over the other or why both exist.  For someone new to the library it might as well be "flibber interface" and "jibberish interface".

With explicit / implicit there is at least some additional information in there about how they the two APIs differ and (maybe) an implication that they have equivalent capabilities.  If people have read the zen of Python it also helps nudge them to using the explicit interface.

**Comment by jklymak** [comment]
In https://matplotlib.org/stable/users/explain/api_interfaces.html#api-interfaces we say "The explicit "Axes" interface" and "The implicit "pyplot" interface".  Maybe just move that into this location as well.

**Comment by zeehio** [comment]
Coming from https://twitter.com/matplotlib/status/1698143656077599077?s=20

The most effective way to avoid confusion between interfaces for users is to say "import X is discouraged, please use import Y instead".

So if there was an `import matplotlib.axes as plt` interface that provided `figure()`, `subplots()`, etc and left `plot()` out, it would be easier for users to tell "this is what matplotlib devs want me to use" vs "this is what I have always used" and then adapt and ask questions about it.

Autocomplete would tell us users what best practice is.

Thanks for all your work!

**Comment by jklymak** [comment]
Yes we have that, but it's not merged yet.  Should be in my opinion. https://github.com/matplotlib/mpl-gui
